src/timestamp_generator.cpp

```diff
@@ -52,24 +61,31 @@ int64_t MonotonicTimestampGenerator::next() {
   }
 }
 
+// This is guaranteed to return a monotonic timestamp. If clock skew is deteced
+// then this method will increment the last timestamp.
 int64_t MonotonicTimestampGenerator::compute_next(int64_t last) {
-  int64_t millis = last / 1000;
-  int64_t counter = last % 1000;
-
-  int64_t now = get_time_since_epoch_ms();
-
-  if (millis >= now) {
-    if (counter == 999) {
-      LOG_WARN("Sub-millisecond counter overflowed, some query timestamps will not be distinct");
-    } else {
-      ++counter;
+  int64_t current = get_time_since_epoch_us();
+
+  if (last >= current) { // There's clock skew
+    // If we execeed our warning threshhold then warn periodically that clock
+    // skew has been detected.
+    if (warning_threshold_us_ >= 0 && last > current + warning_threshold_us_) {
+      int64_t now = get_time_since_epoch_ms();
+      int64_t last_warning = last_warning_.load();
+      if (now > last_warning + warning_interval_ms_ &&
```

Comment:
We just ran into this decision too. We decided that either behavior is reasonable, so we chose not to account for the 1-second offset to match the Java driver's behavior, as we understand it:

https://github.com/datastax/python-driver/pull/681#discussion_r97587655

Open to hearing other opinions if people have strong ones, but I don't see a strong reason to diverge from the Java driver.